src/storage/src/render/upsert.rs

```diff
@@ -435,41 +380,67 @@ where
                         a_ts == b_ts && a_key == b_key
                     });
 
+                    let bincode_opts = types::upsert_bincode_opts();
                     // Upsert the values into `commands_state`, by recording the latest
                     // value (or deletion). These will be synced at the end to the `state`.
+                    //
+                    // Note that we are effectively doing "mini-upsert" here, using
+                    // `command_state`. This "mini-upsert" is seeded with data from `state`, using
+                    // a single `multi_get` above, and the final state is written out into
+                    // `state` using a single `multi_put`. This simplifies `UpsertState`
+                    // implementations, and reduces the number of reads and write we need to do.
+                    //
+                    // This "mini-upsert" technique is actually useful in some `UpsertState`
+                    // implementations, themselves, see `types::rocksdb` for an example. In some
```

Comment:
What does this comment mean? The `UpsertState` implementation should never do this logic and just blindly write what we tell them from here, right?

Reply:
Moved into the consolidated impl in the new commit!